### PR TITLE
feat: Allow configurable PrefetchCount

### LINKED ISSUE #426: Allow configuration of 'PrefetchCount' value while configuring AzureServiceBus messagepump
When we configure an Azure ServiceBus message-pump, it is possible to specify certain configuration options like `MaxConcurrentCalls` etc...
However, it doesn't seem to be possible to configure the `PrefetchCount` setting.

It would be nice if we could configure this.

https://learn.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.servicebusprocessoroptions.prefetchcount?view=azure-dotnet

### PR DESCRIPTION
**Pull request by fgheysels**
Allow to configure the `PrefetchCount` on the ServiceBus message pump.

Closes #426

### CODE REVIEW

File: src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusQueueMessagePumpOptions.cs [hunk 1]
```diff
@@ -13,8 +12,15 @@ public interface IAzureServiceBusQueueMessagePumpOptions
...
-        int? MaxConcurrentCalls { get; set; }
-        
+        int MaxConcurrentCalls { get; set; }
+
+        /// <summary>
+        /// Gets or sets the number of messages that will be eagerly requested from
+        /// Queues or Subscriptions and queued locally, intended to help maximize throughput
+        /// by allowing the processor to receive from a local cache rather than waiting on a service request.
+        /// </summary>
+        public int PrefetchCount { get; set; }
```
Comment (stijnmoreels) [inline src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusQueueMessagePumpOptions.cs:22]:
```suggestion
        int PrefetchCount { get; set; }
```

File: src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusTopicMessagePumpOptions.cs [hunk 2]
```diff
@@ -21,8 +21,15 @@ public interface IAzureServiceBusTopicMessagePumpOptions
...
-        int? MaxConcurrentCalls { get; set; }
-        
+        int MaxConcurrentCalls { get; set; }
+
+        /// <summary>
+        /// Gets or sets the number of messages that will be eagerly requested from
+        /// Queues or Subscriptions and queued locally, intended to help maximize throughput
+        /// by allowing the processor to receive from a local cache rather than waiting on a service request.
+        /// </summary>
+        public int PrefetchCount { get; set; }
```
Comment (stijnmoreels) [inline src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusTopicMessagePumpOptions.cs:31]:
```suggestion
        int PrefetchCount { get; set; }
```

File: src/Arcus.Messaging.Pumps.ServiceBus/Configuration/AzureServiceBusMessagePumpSettings.cs [hunk 3]
```diff
@@ -295,12 +295,13 @@ private ServiceBusProcessorOptions DetermineMessageProcessorOptions()
             {
                 // Assign the configured defaults
                 messageHandlerOptions.AutoCompleteMessages = Options.AutoComplete;
-                messageHandlerOptions.MaxConcurrentCalls = Options.MaxConcurrentCalls ?? messageHandlerOptions.MaxConcurrentCalls;
+                messageHandlerOptions.MaxConcurrentCalls = Options.MaxConcurrentCalls;
+                messageHandlerOptions.PrefetchCount = Options.PrefetchCount;
```
Comment (stijnmoreels) [inline src/Arcus.Messaging.Pumps.ServiceBus/Configuration/AzureServiceBusMessagePumpSettings.cs:299]: Since this is indeed functionality that is out of our hands, it would be a bit too much to test out Microsoft's functionality in integration tests.

File: src/Arcus.Messaging.Pumps.ServiceBus/Configuration/AzureServiceBusMessagePumpSettings.cs [hunk 4]
```diff
@@ -295,12 +295,13 @@ private ServiceBusProcessorOptions DetermineMessageProcessorOptions()
             {
                 // Assign the configured defaults
```
Comment (stijnmoreels) [inline src/Arcus.Messaging.Pumps.ServiceBus/Configuration/AzureServiceBusMessagePumpSettings.cs:296]:
```suggestion
```

File: src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusTopicMessagePumpOptions.cs (hunk 2 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusTopicMessagePumpOptions.cs:31]: Since we're working on a new major version nonetheless, we can create a breaking change here in the interface. 👍

File: src/Arcus.Messaging.Tests.Unit/ServiceBus/AzureServiceBusMessagePumpOptionsTest.cs [hunk 5]
```diff
@@ -43,6 +43,45 @@ public void TopicOptionsMaxConcurrentCalls_ValueIsNegative_ThrowsException()
...
+            // Assert
+            Assert.Equal(validPrefetchCount, options.PrefetchCount);
+        }
+
+        [Fact]
+        public void TopicOptionsPrefetchCount_ValueIsNegative_ThrowsException()
+        {
+            // Arrange
+            var options = new AzureServiceBusMessagePumpOptions();
+            var invalidPrefetchCount = -1;
```
Comment (stijnmoreels) [inline src/Arcus.Messaging.Tests.Unit/ServiceBus/AzureServiceBusMessagePumpOptionsTest.cs:79]: Best that we use `Bogus` to generate our numbers here, instead of hard-code them.

**Review by stijnmoreels — APPROVED** [review]
Thx for this small enhancement, there is a suggestion on the unit test improvement, plus 

- [ ] we also need to update our preview docs in this PR.

https://messaging.arcus-azure.net/Features/message-handling/service-bus

File: src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusQueueMessagePumpOptions.cs (hunk 1 above)
Comment (fgheysels) [reply src/Arcus.Messaging.Pumps.ServiceBus/Configuration/IAzureServiceBusQueueMessagePumpOptions.cs:22]: Woops nice catch!

File: src/Arcus.Messaging.Pumps.ServiceBus/Configuration/AzureServiceBusMessagePumpSettings.cs (hunk 3 above)
Comment (fgheysels) [reply src/Arcus.Messaging.Pumps.ServiceBus/Configuration/AzureServiceBusMessagePumpSettings.cs:298]: Seems like I missed that one.

File: src/Arcus.Messaging.Tests.Unit/ServiceBus/AzureServiceBusMessagePumpOptionsTest.cs (hunk 5 above)
Comment (fgheysels) [reply src/Arcus.Messaging.Tests.Unit/ServiceBus/AzureServiceBusMessagePumpOptionsTest.cs:79]: I can do that, but I'm wondering what would be the advantage of this in this specific situation ?

Comment (stijnmoreels) [reply src/Arcus.Messaging.Tests.Unit/ServiceBus/AzureServiceBusMessagePumpOptionsTest.cs:79]: Removing magic numbers, making the test intent clear (we're not interested in -1, we're interested in negative numbers), and removing edge-case mistakes.

Comment (fgheysels) [reply src/Arcus.Messaging.Tests.Unit/ServiceBus/AzureServiceBusMessagePumpOptionsTest.cs]: Changed this to make use of Bogus, but tbh, I don't see much added value for these specific cases.
Also saw that existing tests were using hardcoded values as well.

**Review by stijnmoreels — APPROVED** [review]

File: docs/preview/02-Features/02-message-handling/01-service-bus.md [hunk 6]
```diff
@@ -425,6 +425,12 @@ public class Startup
                 // (default: null, leading to the defaults of the Azure Service Bus SDK message handler options).
                 options.MaxConcurrentCalls = 5;
 
+                // Specifies the amount of messages that will be eagerly requested during processing.
+                // Setting the PrefetchCount to a value higher then the MaxConcurrentCalls value helps maximizing 
+                // throughput by allowing the MessagePump to receive from a local cache rather then waiting on a
```
Comment (stijnmoreels) [inline docs/preview/02-Features/02-message-handling/01-service-bus.md:430]:
```suggestion
                // throughput by allowing the message pump to receive from a local cache rather then waiting on a 
```

File: docs/preview/02-Features/02-message-handling/01-service-bus.md [hunk 7]
```diff
@@ -475,6 +481,12 @@ public class Startup
...
```
Comment (stijnmoreels) [inline docs/preview/02-Features/02-message-handling/01-service-bus.md:486]:
```suggestion
                // throughput by allowing the message pump to receive from a local cache rather then waiting on a 
```

**Comment by stijnmoreels** [comment]
Thx for this!

File: docs/preview/02-Features/02-message-handling/01-service-bus.md (hunk 7 above)
Comment (fgheysels) [reply docs/preview/02-Features/02-message-handling/01-service-bus.md:486]: Good point.  :)
